### ISSUE TITLE
Cuda Exception: invalid argument

### ISSUE THREAD

**Issue by cpdurham** [AccelerateHS/accelerate-llvm#12]
I had some old code I wanted to run, got this:

```
*** Exception: 
*** Internal error in package accelerate ***
*** Please submit a bug report at https://github.com/AccelerateHS/accelerate/issues
./Data/Array/Accelerate/LLVM/PTX/State.hs:50:9: (unhandled): CUDA Exception: invalid argument
CallStack (from HasCallStack):
  error, called at ./Data/Array/Accelerate/LLVM/PTX/State.hs:50:9 in accelerate-llvm-ptx-1.0.0.0-FE67JRfxrn7CHJHFASzBxC:Data.Array.Accelerate.LLVM.PTX.State
```
with this code:
```haskell
thresholdFunctionFromRight
  :: (A.Ord a, P.Num (Exp a))
  => Exp a
  -> (Exp a, Exp a, Exp a, Exp a, t1, Exp a)
  -> (Exp a, Exp a, Exp a, Exp a, t, Exp a)
  -> (Exp a, Exp a, Exp a, Exp a, t1, Exp a)
thresholdFunctionFromRight
  thresh
  (mi1,l1,c1,r1,li1,ri1)
  (mi2,l2,c2,r2,li2,ri2) =
      (mi3,l3,c3,r3,li3,ri3)
  where
    l3 = l1 + c1 * l2
    c3 = c1 * c2
    r3 = r2 + c2 * r1
    checkRight = A.max 0 . A.min 1 $ A.subtract (thresh-1) r3
    checkCenter = A.max 0 . A.min 1 $ A.subtract (thresh-1) (r1 + l2)
    mi3 = A.max mi1 $ A.max mi2 $ A.max (checkCenter * (ri1+l2)) (checkRight * ri2)
    li3 = li1
    ri3 = ri2

thresholdFunctionFromRightExp
  :: Exp Int
  -> Exp (Int,Int,Int,Int,Int,Int)
  -> Exp (Int,Int,Int,Int,Int,Int)
  -> Exp (Int,Int,Int,Int,Int,Int)
thresholdFunctionFromRightExp thresh e1 e2 =
  let
    v1 = unlift e1 :: (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
    v2 = unlift e2 :: (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
  in
   lift $ thresholdFunctionFromRight thresh v1 v2

convert
  :: Acc (Array DIM1 Int)
  -> Acc (Array DIM1 (Int,Int,Int,Int,Int,Int))
convert arr =
    let
        indicies = A.enumFromN (A.shape arr) 0
    in
      A.zipWith (\a b -> lift6E (0,b,b,b,a,a)) indicies $ arr

lift6E
  :: (Elt a, Elt b, Elt c, Elt d, Elt e, Elt f)
  => (Exp a, Exp b, Exp c, Exp d, Exp e, Exp f)
  -> Exp (a,b,c,d,e,f)
lift6E = lift
```
ran this in ghci:
```
let arr = A.fromList (Z :. 10 :: DIM1) ([0,0,1,1,1,0,1,1,0,0,0,0] :: [Int])
run1 (A.scanl (thresholdFunctionFromRightExp 2) initialV. convert) arr
```
(edit: originally tried scanl, both scanl and scanr have same error)

**Comment by tmcdonell** [comment]
Uh oh, this worked ok for me.

I'm using LLVM-3.9, GHC-8.0.2, AccelerateHS/accelerate-llvm@e9d69e816cac971a4a2fc6edfa86d27365b3c0b5, and device is compute 3.0. What is your configuration?

**Comment by cpdurham** [comment]
Ok, I'm on llvm-3.9, ghc-8.0.2, https://github.com/AccelerateHS/accelerate-llvm/commit/954ffc6e17c6559d9e0d57df77cbb50d98228e97 on Pascal, I'll try on newer llvm-ptx commit when I get a chance. Works in ghci too?

**Comment by cpdurham** [comment]
Got the same error on all the latest accelerate libs, compiled too

**Comment by cpdurham** [comment]
So when i strip it down, and just add elements of the tuples, it fails for tuples of size 6 and 5, and works for 4

**Comment by tmcdonell** [comment]
Ah, so this works:

```haskell

t4 :: Acc (Vector (Int,Int,Int,Int))
t4 = A.zipWith (lift2 f4) xs ys
  where
    xs = use $ fromList (Z:.10) [ (x,x,x,x) | x <- [0..9] ]
    ys = use $ fromList (Z:.10) [ (x,x,x,x) | x <- [0..9] ]

    f4 :: (Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int)
    f4 (x1,x2,x3,x4) (y1,y2,y3,y4) = (x1+y1,x2+y2,x3+y3,x4+y4)
```

But this does not?

```haskell
t5 :: Acc (Vector (Int,Int,Int,Int,Int))
t5 = A.zipWith (lift2 f5) xs ys
  where
    xs = use $ fromList (Z:.10) [ (x,x,x,x,x) | x <- [0..9] ]
    ys = use $ fromList (Z:.10) [ (x,x,x,x,x) | x <- [0..9] ]

    f5 :: (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
    f5 (x1,x2,x3,x4,x5) (y1,y2,y3,y4,y5) = (x1+y1,x2+y2,x3+y3,x4+y4,x5+y5)
```

Yes, this is working in GHCi for me.

**Comment by cpdurham** [comment]
Those both work, but when i do this 
```
t4 :: Acc (Vector (Int,Int,Int,Int))
t4 = A.scanl1 (lift2 f4) xs --A.zipWith (lift2 f4) xs ys
  where
    xs = use $ fromList (Z:.10) [ (x,x,x,x) | x <- [0..9] ]
--    ys = use $ fromList (Z:.10) [ (x,x,x,x) | x <- [0..9] ]

    f4 :: (Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int)
    f4 (x1,x2,x3,x4) (y1,y2,y3,y4) = (x1+y1,x2+y2,x3+y3,x4+y4)

t5 :: Acc (Vector (Int,Int,Int,Int,Int))
t5 = A.scanl1 (lift2 f5) xs --A.zipWith (lift2 f5) xs ys
  where
    xs = use $ fromList (Z:.10) [ (x,x,x,x,x) | x <- [0..9] ]
--    ys = use $ fromList (Z:.10) [ (x,x,x,x,x) | x <- [0..9] ]

    f5 :: (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
       -> (Exp Int, Exp Int, Exp Int, Exp Int, Exp Int)
    f5 (x1,x2,x3,x4,x5) (y1,y2,y3,y4,y5) = (x1+y1,x2+y2,x3+y3,x4+y4,x5+y5)
```
t4 works, but t5 gives me that same error

**Comment by cpdurham** [comment]
Playing around, I traded out my cards and a maxwell card had no problem with those on the same configuration, so it's definitely the pascal card to blame

**Comment by tmcdonell** [comment]
Would you mind trying again with HEAD, which now uses LLVM-4.0? (on the off chance it was a problem with 3.9 on the newest hardware).

**Comment by cpdurham** [comment]
Just tried it, still getting the same error

**Comment by tmcdonell** [comment]
This was a bug in the occupancy calculator, requesting more shared memory than allowed. Fixed with [cuda-0.7.5.3](https://hackage.haskell.org/package/cuda-0.7.5.3).

Sorry this took so long!
